### PR TITLE
Round environment

### PR DESCRIPTION
**Pull request by vagechirkov**
example

### CODE REVIEW

**Comment by vagechirkov** [comment]
I'm wondering how to make better videos 😅

**Comment by vagechirkov** [comment]
I think the problem with the workflow is because of this:
image

I'm not sure how we can fix this... 🤔

@mezdahun

**Comment by mezdahun** [comment]
@vagechirkov that arena looks awesome! :star_struck: 

regarding the Docker CLI automation you could try to add the missing credentials to your github account (in case you have DockerHub credentials). If that doesn't work, from now on you could make indirect PRs to first just feature branches in the main repo instead of direct PR from fork feature branch to develop in original repo. This automation is only executed on pushes and PRs to origin/develop and is testing installable dependencies on docker environment. Currently if you change e.g. `setup.py` or `Dockerfile` in a way that it is messing up the docker image there is no way to catch it before merging to develop as the test must be manually bypassed.

**Comment by vagechirkov** [comment]
> @vagechirkov that arena looks awesome! 🤩
> 
> regarding the Docker CLI automation you could try to add the missing credentials to your github account (in case you have DockerHub credentials). If that doesn't work, from now on you could make indirect PRs to first just feature branches in the main repo instead of direct PR from fork feature branch to develop in original repo. This automation is only executed on pushes and PRs to origin/develop and is testing installable dependencies on docker environment. Currently if you change e.g. `setup.py` or `Dockerfile` in a way that it is messing up the docker image there is no way to catch it before merging to develop as the test must be manually bypassed.

Hmm, I see. I have the credentials in my fork and the workflow works fine. For example, I enabled workflow for `round-environment` branch and Docker Image CI works fine https://github.com/vagechirkov/ABM/tree/round-environment but not in the PR from the fork (probably because of the security issues).

image

**Comment by vagechirkov** [comment]
We can run Docker CI on all branches or all branches with the specific pattern (e.g. "develop/**") but not on the pull request. I don't think it's possible to get secrets from the fork to run a workflow for PR. @mezdahun what do you think? In principle, we can first push into another branch and then merge into the `develop` branch, but I'm not sure if that's an optimal solution 🤔.

**Comment by mezdahun** [comment]
I see! In that case the preferred workflow would be to not PR directly to develop in the original repo but rather to a feature branch first. In that PR general tests are run. Then I PR in the original repo the feature branch to develop and check Docker CLI. Would that work for you?

**Comment by mezdahun** [comment]
@vagechirkov What do you mean by it's not the optimal solution?

**Comment by mezdahun** [comment]
@vagechirkov https://github.com/danger/danger-js/issues/918#issuecomment-645814203 I guess something like this would be nicer, but it wouldn't solve the problem I mentioned before, namely we merge without checking if the resulting docker image is healthy

**Comment by mezdahun** [comment]
Another option is to do direct PRs from your fork to develop but I will check the result of Docker CLI test on your feature branch before bypassing the check on the PR.

**Comment by vagechirkov** [comment]
> @vagechirkov What do you mean by it's not the optimal solution?

Well, in principle I think it's ok to go through the feature branch before merging commits with the main branch. It's just a bit redundant. 🙃

**Comment by vagechirkov** [comment]
I'm wondering why this is necessary to run Docker CI before merging changes to the repo. I think we can run tests of the code with `pytest` in the different workflow and use the Docker CI workflow just to push a new image to Docker Hub. 🤔

**Comment by mezdahun** [comment]
The idea behind is I guess that we want to keep our image on DockerHub always in sync with develop (as this will serve as the environment for HPC simulations). In case something is merged that messes up the image they get out of sync and you have to fix that on a different branch again (after merge). This is why already on the PR side it is checked. What we could do here due to the credential issues from fork is to instead separate the docker build and the push in 2 different github action files. This way the docker build would run on the PR side allowing us to be sure that the resulting image would be okay. The one with push to DockerHub would then be only activated upon/after merge. This would be also relatively easy I think. Which one would be your preferred option?

1. PR to feature branches instead of develop
2. Check docker result on remote branch then bypass check at PR to develop
3. Separate Docker check to 2 checks (1 on PR only build, 2 on merge build and push)

**Comment by mezdahun** [comment]
Or you meant the same thing?

**Comment by vagechirkov** [comment]
> Or you meant the same thing?

Yes indeed, I like the third option! Let me try to build the image without login :)

**Comment by vagechirkov** [comment]
Cool! I think this should work! It makes a lot of sense to only push the Docker image to Docker Hub after merging PR.

**Comment by mezdahun** [comment]
Awesome, I will do a review later and we can merge this!

File: abm/projects/cooperative_signaling/cs_agent/cs_agent.py
Comment (mezdahun) [inline abm/projects/cooperative_signaling/cs_agent/cs_agent.py]: We are changing the position of the agents twice in a single time step. Once in `self.update()` and once here.

File: abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py
Comment (mezdahun) [inline abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py:67]: I understand this fixes the behavior, but I am wondering why it is necessary.

File: abm/projects/cooperative_signaling/cs_agent/tests/test_cs_supcalc.py
Comment (mezdahun) [inline abm/projects/cooperative_signaling/cs_agent/tests/test_cs_supcalc.py:15]: for me random walk test fails can you please check?

Comment (mezdahun) [reply abm/projects/cooperative_signaling/cs_agent/tests/test_cs_supcalc.py:15]: Ah it is because the env parameters, we might want to mock this so that a new local env file won't influence the test results

File: abm/projects/cooperative_signaling/cs_environment/cs_resource.py
Comment (mezdahun) [inline abm/projects/cooperative_signaling/cs_environment/cs_resource.py]: magic number

File: abm/projects/cooperative_signaling/cs_simulation/cs_sims.py
Comment (mezdahun) [inline abm/projects/cooperative_signaling/cs_simulation/cs_sims.py:171]: for me this still yield new agents outside of the circular arena when I am increasing the agent number in playground tool, this is probably in cs_isims.py that we need to ovverride

**Review by mezdahun — CHANGES_REQUESTED** [review]
Awesome! A few things which we should fix before merging is:
- fix placing new agents into circular arena when playground tool is used
- update the position of the agents only once per time step
- maybe comment a bit on why that check of new angle is needed. i.e. how it can happen that the calculation yields a new angle that points outside of the circle? 
- remove magic numbers
- BUG: if the resource patch is small and close to the border while user increases the patch radius (detection range) the border gets stuck at the wall.
- Discuss: seemingly the resulting angles are very sharp. We could think about a recursive turning method inside? Maybe also not necessary if this approach works.

File: abm/projects/cooperative_signaling/cs_agent/tests/test_cs_supcalc.py
Comment (vagechirkov) [reply abm/projects/cooperative_signaling/cs_agent/tests/test_cs_supcalc.py:15]: This is strange because all the tests in the workflow passed 🤔

File: abm/projects/cooperative_signaling/cs_agent/cs_agent.py
Comment (vagechirkov) [reply abm/projects/cooperative_signaling/cs_agent/cs_agent.py]: What do you think about this solution? b990a130417fe073f702a644030a5d59986bc8f9

File: abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py
Comment (vagechirkov) [reply abm/projects/cooperative_signaling/cs_agent/cs_supcalc.py:67]: I would guess it has something to do with the curvature of the boundary and the step size per update 😅

File: abm/projects/cooperative_signaling/cs_environment/cs_resource.py
Comment (vagechirkov) [reply abm/projects/cooperative_signaling/cs_environment/cs_resource.py]: Should be better now 768941d65dad9ac86d0e460f447229bc9408ac42

File: abm/projects/cooperative_signaling/cs_simulation/cs_sims.py
Comment (vagechirkov) [reply abm/projects/cooperative_signaling/cs_simulation/cs_sims.py:171]: Nice, I missed that bug! This should solve it: 12a35dca8d530c28b595b8ac11eee8898ee10eb2

File: abm/projects/cooperative_signaling/cs_agent/tests/test_cs_supcalc.py
Comment (mezdahun) [reply abm/projects/cooperative_signaling/cs_agent/tests/test_cs_supcalc.py:15]: If you replace the .env file (i.e. you are after running the playground tool) the tests won't pass. Or do they pass for you even with a new .env file?

Comment (vagechirkov) [reply abm/projects/cooperative_signaling/cs_agent/tests/test_cs_supcalc.py:15]: I never had any problems with tests before or after running the playground 😅🤔

**Comment by vagechirkov** [comment]
@mezdahun Please let me know if there is anything else we need to fix before merging this PR. I'm not sure why tests are failing for you, but I think we're good as tests are passing in the GitHub workflow 🙂

**Comment by mezdahun** [comment]
In the github workflow tests pass because it uses the same prototype .env file as always. Once you change your env file locally and run pytest they will fail. (as soon as you change env parameters that your tests use via the function to be tested)

Usually this should be mocked, as in env parameters are fixed for the given run of a test case. I guess this si our of scope here, it is enough if we make the tests for the prototype env file.

Should we look into this bug with increasing the detection range of the resource around the wall?

To reproduce: Stop the simulation (with space) and move the resource patch near the wall. Increase the detection range so that the patch is now out of the arena partially. Resume the simulation (with space).

Expected behavior: the resource is getting relocated and bouncing back from wall of arena

Behavior: The patch gets stuck and won't move anymore

**Comment by vagechirkov** [comment]
> In the github workflow tests pass because it uses the same prototype .env file as always. Once you change your env file locally and run pytest they will fail. (as soon as you change env parameters that your tests use via the function to be tested)
> 
> Usually this should be mocked, as in env parameters are fixed for the given run of a test case. I guess this si our of scope here, it is enough if we make the tests for the prototype env file.
> 
> Should we look into this bug with increasing the detection range of the resource around the wall?
> 
> To reproduce: Stop the simulation (with space) and move the resource patch near the wall. Increase the detection range so that the patch is now out of the arena partially. Resume the simulation (with space).
> 
> Expected behavior: the resource is getting relocated and bouncing back from wall of arena
> 
> Behavior: The patch gets stuck and won't move anymore

Could you fix this .env file problem? Unfortunately, I cannot reproduce it in my setup. I will fix the relocation of recourse inside the arena 🙂

**Comment by mezdahun** [comment]
> > In the github workflow tests pass because it uses the same prototype .env file as always. Once you change your env file locally and run pytest they will fail. (as soon as you change env parameters that your tests use via the function to be tested)
> > Usually this should be mocked, as in env parameters are fixed for the given run of a test case. I guess this si our of scope here, it is enough if we make the tests for the prototype env file.
> > Should we look into this bug with increasing the detection range of the resource around the wall?
> > To reproduce: Stop the simulation (with space) and move the resource patch near the wall. Increase the detection range so that the patch is now out of the arena partially. Resume the simulation (with space).
> > Expected behavior: the resource is getting relocated and bouncing back from wall of arena
> > Behavior: The patch gets stuck and won't move anymore
> 
> Could you fix this .env file problem? Unfortunately, I cannot reproduce it in my setup. I will fix the relocation of recourse inside the arena slightly_smiling_face

https://github.com/scioip34/ABM/pull/48/commits/09f67da63b95f56379bbd4666e2291f7720e4f6f should fix this. I mocked the movementparams submodule's fixed variables that is used by the random walk function if no parameters are passed to it. In that case (when function parameters are None) it fetched the movementparams submodules corresponding attributes (velocity, min and max theta). These are read from the root .env file upon startup of the stack.

**Review by mezdahun — APPROVED** [review]
:rocket:

**Comment by vagechirkov** [comment]
LGTM, let's merge! However, I don't have a merge button 😂

**Comment by mezdahun** [comment]
Oh sorry I thought you had one

**Comment by vagechirkov** [comment]
> Oh sorry I thought you had one

This is probably set in the protected branch settings 🤔

**Pull request merged**
